sources/org.osbuild.curl

```diff
@@ -93,6 +98,135 @@ SCHEMA = """
+    # --write-out "%{json}" was added in 7.70
+    # --write-out "%{exitcode} is 7.75
+    if int(major) == 7 and int(minor) >= 75:
+        return True
+    return False
+
+
+def _quote_url(url: str) -> str:
+    purl = urllib.parse.urlparse(url)
+    path = urllib.parse.quote(purl.path)
+    quoted = purl._replace(path=path)
+    return quoted.geturl()
+
+
+def gen_curl_download_config(config_path: pathlib.Path, chksum_desc_tuple: List[Tuple[str, Dict]]):
+    with open(config_path, "w", encoding="utf8") as fp:
+        # global options
+        fp.write(textwrap.dedent(f"""\
+        header = "User-Agent: osbuild (Linux.{platform.machine()}; https://osbuild.org/)"
+        silent
+        speed-limit = 1000
+        connect-timeout = 30
+        fail
+        location
+        """))
+        proxy = os.getenv("OSBUILD_SOURCES_CURL_PROXY")
+        if proxy:
+            fp.write(f'proxy = "{proxy}"\n')
+        fp.write("\n")
+        # per url options
+        for checksum, desc in chksum_desc_tuple:
+            url = _quote_url(desc.get("url"))
+            fp.write(f'url = "{url}"\n')
+            fp.write(f'output = "{checksum}"\n')
+            secrets = desc.get("secrets")
+            if secrets:
+                ssl_ca_cert = secrets.get('ssl_ca_cert')
+                if ssl_ca_cert:
+                    fp.write(f'cacert = "{ssl_ca_cert}"\n')
+                ssl_client_cert = secrets.get('ssl_client_cert')
+                if ssl_client_cert:
+                    fp.write(f'cert = "{ssl_client_cert}"\n')
+                ssl_client_key = secrets.get('ssl_client_key')
+                if ssl_client_key:
+                    fp.write(f'key = "{ssl_client_key}"\n')
+            insecure = desc.get("insecure")
+            if insecure:
+                fp.write('insecure\n')
+            else:
+                fp.write('no-insecure\n')
+            fp.write("\n")
+
+
+def try_parse_curl_json_line(line):
+    line = line.strip()
+    if line:
+        try:
+            return json.loads(line)
+        except json.decoder.JSONDecodeError:
+            print(f"WARNING: cannot decode {line}")
+    return None
+
+
+def fetch_many_new_curl(tmpdir, targetdir, dl_pairs):
+    curl_config_path = f"{tmpdir}/curl-config.txt"
+    gen_curl_download_config(curl_config_path, dl_pairs)
+    curl_command = [
+        "curl",
+        "--config", curl_config_path,
+        # this adds a bunch of noise but might be nice for debug?
+        # "--show-error",
+        "--parallel",
+        # this will write out a json record for each finished download
+        "--write-out", "%{json}\n",
+    ]
+    with contextlib.ExitStack() as cm:
+        curl_p = subprocess.Popen(curl_command, encoding="utf-8", cwd=tmpdir, stdout=subprocess.PIPE)
+        # ensure that curl is killed even if an unexpected exit happens
+        cm.callback(curl_p.kill)
+        while curl_p.poll() is None:
+            dl_details = try_parse_curl_json_line(curl_p.stdout.readline())
```

Comment:
This is potentially racy, we need to double check that we have "drained" stdout before curl_p.poll() exits

Reply:
The race is fixed.